sensor/kubernetes/listener/resources/virtualmachine/dispatcher/virtualmachineinstances.go

```diff
@@ -0,0 +1,99 @@
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/sensor/kubernetes/eventpipeline/component"
+	"github.com/stackrox/rox/sensor/kubernetes/listener/resources/virtualmachine/store"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeVirtV1 "kubevirt.io/api/core/v1"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+//go:generate mockgen-wrapper
+type virtualMachineStore interface {
+	Get(uid string) *store.VirtualMachineInfo
+	AddOrUpdateVirtualMachine(vm *store.VirtualMachineInfo)
+	AddOrUpdateVirtualMachineInstance(uid, namespace string, vsockCID *uint32, isRunning bool)
+	RemoveVirtualMachine(uid string)
+	RemoveVirtualMachineInstance(uid string)
+}
+
+type VirtualMachineInstanceDispatcher struct {
+	clusterID string
+	store     virtualMachineStore
+}
+
+func NewVirtualMachineInstanceDispatcher(clusterID string, store virtualMachineStore) *VirtualMachineInstanceDispatcher {
+	return &VirtualMachineInstanceDispatcher{
+		clusterID: clusterID,
+		store:     store,
+	}
+}
+
+func (d *VirtualMachineInstanceDispatcher) ProcessEvent(
+	obj interface{},
+	_ interface{},
+	action central.ResourceAction,
+) *component.ResourceEvent {
+	unstructuredObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		log.Errorf("not of type 'Unstructured': %T", obj)
+		return nil
+	}
+	virtualMachineInstance := &kubeVirtV1.VirtualMachineInstance{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObj.Object, virtualMachineInstance); err != nil {
+		log.Errorf("unable to convert 'Unstructured' to 'VirtualMachineInstance': %v", err)
+		return nil
+	}
+	if virtualMachineInstance.GetUID() == "" {
+		log.Errorf("convertion from unstructured failed: %v", obj)
+		return nil
+	}
+	if len(virtualMachineInstance.GetOwnerReferences()) != 1 {
+		log.Errorf("virtual machine instance with no owner reference %v", virtualMachineInstance.GetOwnerReferences())
+		return nil
+	}
+	vmUID := string(virtualMachineInstance.GetOwnerReferences()[0].UID)
+	ownerEventReceived := d.store.Get(vmUID) != nil
+	if action == central.ResourceAction_REMOVE_RESOURCE {
+		d.store.RemoveVirtualMachineInstance(vmUID)
+	} else {
+		isRunning := virtualMachineInstance.Status.Phase == kubeVirtV1.Running
+		vsock := virtualMachineInstance.Status.VSOCKCID
+		d.store.AddOrUpdateVirtualMachineInstance(vmUID, virtualMachineInstance.Namespace, vsock, isRunning)
+	}
+
+	// Do not send UPDATE events if:
+	// - we are syncing resources
+	// - we did not receive the VirtualMachine associated with this instance yet
+	if action == central.ResourceAction_SYNC_RESOURCE || !ownerEventReceived {
+		return nil
+	}
+
+	vmInfo := d.store.Get(vmUID)
+	if vmInfo == nil {
+		log.Error("store does not contain the owner reference to this instance")
+		return nil
+	}
+
+	// VirtualMachineInstances are not tracked by Central.
+	// Send an UPDATE event of the VirtualMachine associated with this instance
```

Comment:
So VIrtualMachineInstance is only a Sensor thing? I think, I am missing some knowledge about this feat as this is a bit surprising for me. Could you maybe briefly share why track VMI in Sensor only?

Reply:
Okay, so if we have, for example 5 running VMIs, then Central would only experience this through `VirtualMachine`, right?  I guess that all those 5 VMIs would need to be identical for the purpose of vuln scanning. I thought that this is difficult to guarantee as the VMIs would need to be immutable. Do we have such guarantee?